test/unittests/utils/test_connection_live.py

```diff
@@ -22,3 +227,118 @@ def test_anonymize_different_lengths() -> None:
     assert con._anonymize({"token": "uk7m20-8gqn8ir7e30"}) == {"token": "uk7m2[+13]"}
     assert con._anonymize({"token": "uk7m2"}) == {"token": "*****"}
     assert con._anonymize({"token": "u"}) == {"token": "*"}
+
+
+def test_try_network_action() -> None:
+    con = ConnectionLive("http://example.com/")
+    response_expected = Mock(status_code=200)
+    con.session.request = Mock(return_value=response_expected)
+    con._log_request = Mock()
+    con._log_response = Mock()
+    params = RequestParameters(method="GET", url="http://example.com/", timeout=1)
+    response = con._try_network_action(params)
+    assert response == response_expected
+    con._log_request.assert_called_once_with(params)
+    con.session.request.assert_called_once_with(**params.as_kwargs())
+    con._log_response.assert_called_once_with(response_expected)
+
+
+@patch.dict("os.environ", {})
+def test_try_network_action_timeout_error() -> None:
+    con = ConnectionLive("http://example.com/")
+    responses = (TimeoutError(), TimeoutError(), ReadTimeout(), ReadTimeout(), Mock(status_code=200))
+    session_mock = SessionMock(responses)
+    con.session = session_mock  # type: ignore[assignment]
+    con._log_request = Mock()
+    con._log_response = Mock()
+    params = RequestParameters(method="GET", url="http://example.com/", timeout=1)
+    with patch("dsp_tools.utils.connection_live.time.sleep") as sleep_mock:
+        response = con._try_network_action(params)
+        assert [x.args[0] for x in sleep_mock.call_args_list] == [1, 2, 4, 8]
+    assert [x.args[0] for x in con._log_request.call_args_list] == [params] * len(session_mock.responses)
+    con._log_response.assert_called_once_with(session_mock.responses[-1])
+    assert response == session_mock.responses[-1]
+
+
+@patch.dict("os.environ", {})
+def test_try_network_action_connection_error() -> None:
+    con = ConnectionLive("http://example.com/")
+    responses = (ConnectionError(), ConnectionError(), RequestException(), Mock(status_code=200))
+    session_mock = SessionMock(responses)
+    con.session = session_mock  # type: ignore[assignment]
+    con._log_request = Mock()
+    con._log_response = Mock()
+    con._renew_session = Mock()
+    params = RequestParameters(method="POST", url="http://example.com/", timeout=1)
+    with patch("dsp_tools.utils.connection_live.time.sleep") as sleep_mock:
+        response = con._try_network_action(params)
+        assert [x.args[0] for x in sleep_mock.call_args_list] == [1, 2, 4]
+    assert con._renew_session.call_count == len(session_mock.responses) - 1
+    assert [x.args[0] for x in con._log_request.call_args_list] == [params] * len(session_mock.responses)
+    con._log_response.assert_called_once_with(session_mock.responses[-1])
+    assert response == session_mock.responses[-1]
+
+
+@patch.dict("os.environ", {"DSP_TOOLS_TESTING": "false"})
```

Comment:
Strange. In the 2 tests above, `@patch.dict("os.environ", {})` works fine, but in this one, I need this (rather unrealistic) line for it to work.